net processing: Only support version 2 compact blocks

jnewbery opened the pull request:
Compact blocks are used for efficient relay of blocks, either through High Bandwidth or Low Bandwidth mode. See [BIP 152](https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki) for full details.

For compact block relay to work, the receiver must have a mempool containing transactions which are likely to be included in the block. The receiver uses these transactions to reconstruct the block from the short transaction ids included in the `cmpctblock` message. Compact blocks are therefore only useful for relaying blocks at or near the tip of the block chain. For older blocks, the recipient won't have the transactions in their mempool and so would need to request them using a `getblocktxn` message. In such cases, just requesting the full block is more efficient.

BIP 152 supports two versions: version 1 (without witnesses) and version 2 (with witnesses). Version 2 is required to reconstruct segwit blocks. Segwit was activated in August 2017, and providing non-witness blocks to peers is no longer useful. Since the witnesses are not included, the peer would not be able to fully validate all the consensus rules on the provided block.

Therefore, stop supporting version 1 compact blocks. Ignore `sendcmpct` messages with version=1, and don't advertise support by sending `sendcmpct` with version=1. Only send `sendcmpct` to peers with `NODE_WITNESS`. Respond to all requests for compact blocks or blocktxns with witness-serialized blocks and transactions.

MarcoFalke commented:
> Segwit was activated in August 2017, and providing non-witness blocks to peers is no longer useful

I think there may be some edge-case scenarios where a peer skips the download of witnesses because they are never read (due to an assumevalid setting), but that has nothing to do with compact blocks. So:

Concept ACK on removing code that doesn't serve any use case in practice.

sipa commented:
Concept ACK on removing non-witness cmpctblock support. IIRC its only purpose is serving Bitcoin Core v0.13.0 nodes (0.12.x didn't have compact blocks; 0.13.1 added segwit activation parameters).

The first commit here seems to mix removal of functionality with some refactoring. Is it possible to separate those more cleanly into separate commits?

jnewbery commented:
> The first commit here seems to mix removal of functionality with some refactoring. Is it possible to separate those more cleanly into separate commits?

Yes, I'll see what I can do to split it up a bit more.

practicalswift commented:
Concept ACK

ariard commented on src/net_processing.cpp:
```diff
@@ -534,33 +519,31 @@ static void MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid, CConnman& connma
-                lNodesAnnouncingHeaderAndIDs.pop_front();
-            }
-            connman.PushMessage(pfrom, CNetMsgMaker(pfrom->GetCommonVersion()).Make(NetMsgType::SENDCMPCT, /*fAnnounceUsingCMPCTBLOCK=*/true, nCMPCTBLOCKVersion));
+            connman.PushMessage(pfrom, CNetMsgMaker(pfrom->GetCommonVersion()).Make(NetMsgType::SENDCMPCT,  /* high_bandwidth= */ true, /* version= */ uint64_t{2}));
```
Maybe a static const int CMPCTBLOCK_VERSION? It sounds non-ambiguous, after this PR we only support one version.

jnewbery replied:
Good idea! I'll add.

jnewbery replied:
Done

ariard reviewed (commented):
Concept ACK, thanks for cleaning a bit compact blocks code.

jonatack commented:
Concept ACK, looks like a good simplification. Worth reviewing the diffs with `-w` in some parts.

jnewbery committed:
[net] Stop testing version 1 compact blocks.

Support for version 1 is removed in the following commits.

jnewbery committed:
[net processing] Only advertise support for version 2 compact blocks

Subsequent commits will remove support.

jnewbery committed:
[net processing] Only accept `sendcmpct` with version=2

Subsequent commits will remove support for other versions of compact blocks.

Add a test that a received `sendcmpct` message with version = 1 is
ignored.

jnewbery committed:
[net processing] Simplify `sendcmpct` processing

nCMPCTBLOCKVersion must always be 2 when processing.

jnewbery committed:
[net processing] Remove fSupportsDesiredCmpctVersion

It is now completely redundant with fProvidesHeadersAndIDs.

jnewbery committed:
[net processing] fPreferHeaderAndIDs implies fProvidesHeaderAndIDs

Remove all if(fProvidesHeaderAndIDs) conditionals inside
if(fPreferHeaderAndIDs) conditionals.

jnewbery committed:
[net processing] Remove fWantsCmpctWitness

It is now completely redundant with fProvidesHeadersAndIDs.

jnewbery committed:
[net processing] Tidy up `sendcmpct` processing

- use better local variable names
- drop unnecessary if statements

jnewbery committed:
[net processing] Rename CNodeState compact block members

fPreferHeaderAndIDs -> m_requested_hb_cmpctblocks
fProvidesHeaderAndIDs -> m_provides_cmpctblocks

jnewbery committed:
[net processing] Only relay blocks by cmpctblock and cache for fast relay if segwit is enabled

This introduces an early exit in PeerManagerImpl::NewPoWValidBlock() if
segwit has not been activated for the block. This means that we won't cache the
block/compact block for fast relay and won't relay the cmpctblock
immediately to peers that have requested hb compact blocks. This is fine
because any block where segwit is not yet activated is buried deep in
the chain, and so compact block relay will not be effective.

It's ok not to cache the block/compact block for fast relay for the same
reason - the block must be very deeply buried in the block chain.

ProcessBlockAvailability() also won't get called for all nodes. This is
also fine, since that function only updates hashLastUnknownBlock
and pindexBestKnownBlock, and is called early in every SendMessages()
call.

jnewbery committed:
[net processing] Remove redundant nodestate->m_sendcmpct check in MaybeSetPeerAsAnnouncingHeaderAndIDs()

jnewbery commented:
Rebased and broken up the first commit.

I've also dropped the last two commits (_Clean up PeerManager::BlockChecked()_ and _Clean up MaybeSetPeerAsAnnouncingHeaderAndIDs()_). There's already enough happening in this PR.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -535,7 +528,7 @@ static void MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid, CConnman& connma
     AssertLockHeld(cs_main);
     CNodeState* nodestate = State(nodeid);
-    if (!nodestate || !nodestate->fSupportsDesiredCmpctVersion) {
+    if (!nodestate || !nodestate->fProvidesHeadersAndIDs) {
```
:flushed: Thanks!

jnewbery replied:
fixed

felipsoarez reviewed (commented):
utACK

fjahr commented:
Concept ACK

jonatack reviewed (commented):
Reviewed first three commits up to 55c9231dc then started getting bogged down in the granular ones going over some of the same parts repeatedly and with the renamings; skim-reviewed them for now and will need to return.  All of the commits build cleanly rebased to master with `p2p_compactblocks.py` passing on each.

ariard commented on src/net_processing.cpp:
```diff
@@ -317,12 +317,10 @@ struct CNodeState {
-      */
-    bool fProvidesHeaderAndIDs;
+    /** Whether this peer wants cmpctblocks (when possible) for block announcements. */
+    bool m_sendcmpct_hb{false};
```
IMO `m_prefercmcpt`  better, otherwise name collusion is too close.

jnewbery replied:
`m_prefercmpct` is not as precise. The peer might 'prefer' to fetch blocks using getdata(CMPCT), which is not what this means. `m_sendcmpct_hb` unambiguously means "send blocks to this peer using BIP 152 high bandwidth mode".

theuni replied:
Could you add the "high bandwidth" part of that to the comment so the _hb makes more sense?

ariard reviewed (commented):
I think you should also modify `doc/bips.md` to mention we don't support BIP 152 version 1 anymore.

jnewbery commented on test/functional/p2p_compactblocks.py:
```diff
@@ -141,21 +138,16 @@ def make_utxos(self):
     #   made with compact blocks.
     # - If sendcmpct is then sent with boolean 1, then new block announcements
     #   are made with compact blocks.
-    # If old_node is passed in, request compact blocks with version=preferred-1
```
Good catch. Fixed!

jnewbery commented on src/net_processing.cpp:
```diff
@@ -146,6 +146,8 @@ static constexpr uint32_t MAX_GETCFILTERS_SIZE = 1000;
 /** the maximum percentage of addresses from our addrman to return in response to a getaddr message. */
 static constexpr size_t MAX_PCT_ADDR_TO_SEND = 23;
+/** The compactblocks version we support. See BIP 152. */
+static constexpr uint64_t CMPCTBLOCKS_VERSION = 2;
```
The version field needs to be 8 bytes (uint64_t). If we change this const to a uint8_t, then the message serialized here:

```
            m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::SENDCMPCT, /* high_bandwidth= */ false, /* version= */ CMPCTBLOCKS_VERSION));
```

would have a payload size of 2 (1 byte for high_bandwidth, and 1 byte for version). That would fail to be parsed by other clients.

sipa replied:
That's a bit brittle in any case. Perhaps put `uint64_t{CMPCTBLOCKS_VERSION}` there instead?

jonatack replied:
"followed by a 8-byte integer" -> I misread, thanks

jnewbery replied:
or perhaps:

```
            static_assert(sizeof(CMPCTBLOCKS_VERSION) == sizeof(uint64_t), "sendcmpct version field must be 8 bytes");
            m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::SENDCMPCT, /* high_bandwidth= */ false, /* version= */ CMPCTBLOCKS_VERSION));
```

(why check at runtime what you can test at compile time?)

MarcoFalke replied:
uint64_t{CMPCTBLOCKS_VERSION}` is a compile time check ;)

jnewbery replied:
I've added the static_assert.

jnewbery replied:
> uint64_t{CMPCTBLOCKS_VERSION} is a compile time check ;)

`uint64_t{CMPCTBLOCKS_VERSION}` initializes a new `uint64_t` from `CMPCTBLOCKS_VERSION` at runtime, no? But I guess the compiler will optimize that away?

In any case, I've removed the static assert and replaced with `uint64_t{CMPCTBLOCKS_VERSION}`.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2496,18 +2473,14 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
-        if (pfrom.GetCommonVersion() >= SHORT_IDS_BLOCKS_VERSION) {
-            // Tell our peer we are willing to provide version 1 or 2 cmpctblocks
+        if (pfrom.GetCommonVersion() >= SHORT_IDS_BLOCKS_VERSION &&
+            WITH_LOCK(cs_main, {return State(pfrom.GetId())->fHaveWitness;})) {
```
> can we avoid locking the mutex

Very nice observation, although I think this is ok for now. We take cs_main when processing almost all messages in net_processing (although I agree with you that it'd be slightly nicer if we didn't have to). In the new code, we're only sending `sendcmpct` if the peer has told us that they support witness serialization. That's indicated by fHaveWitness or pfrom.nServices (pfrom.GetLocalServices() is something different - it's the services we offered to the peer).

fHaveWitness will move into Peer very soon as part of #19398, at which point it'll no longer be guarded by cs_main.

> (and order dependance on fHaveWitness being set)

`fHaveWitness` can only be set once in version handling, and this code in verack handling can only be hit after that.

jnewbery replied:
> Maybe we can rename SHORT_IDS_BLOCK_VERSION to SENDCMPCT_VERSION

I do have a branch somewhere that does that, but it's out of scope for this PR.

jnewbery replied:
> I do have a branch somewhere that does that, but it's out of scope for this PR.

Here you go: https://github.com/jnewbery/bitcoin/commit/82372335cc3cf8fa238678629dda6a2f465e6005. Feel free to PR that commit if you think it's worth it.

theuni replied:
Agree with @jonatack about taking from `(pfrom.nServices & NODE_WITNESS)` rather than `State(pfrom.GetId())->fHaveWitness;})` just because it's easier to not have to worry about where the latter comes from.

Understood that it's moot as it's changing soon anyway, so not a big deal.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2496,18 +2473,14 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
-                m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::SENDCMPCT, fAnnounceUsingCMPCTBLOCK, nCMPCTBLOCKVersion));
-            nCMPCTBLOCKVersion = 1;
-            m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::SENDCMPCT, fAnnounceUsingCMPCTBLOCK, nCMPCTBLOCKVersion));
+            m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::SENDCMPCT, /* high_bandwidth= */ false, /* version= */ CMPCTBLOCKS_VERSION));
```
Good suggestion. Done.

jnewbery commented on test/functional/p2p_compactblocks.py:
```diff
@@ -141,19 +138,16 @@ def make_utxos(self):
-            # Check that the first version received is the preferred one
-            assert_equal(test_node.last_sendcmpct[0].version, preferred_version)
+            # Check that the first version received is version 2
+            assert_equal(test_node.last_sendcmpct[0].version, 2)
```
I'm not sure if that adds to the readability here. Leaving for now.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2496,18 +2498,14 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+        if (pfrom.GetCommonVersion() >= SHORT_IDS_BLOCKS_VERSION &&
+            WITH_LOCK(cs_main, {return State(pfrom.GetId())->fHaveWitness;})) {
+            // Tell our peer we are willing to provide version 2 cmpctblocks.
             // However, we do not request new block announcements using
```
This comment is saying that we don't request hb mode from the peer. We only add the peer to one of our three hb peers once it provides us with a block.

In any case, this comment is unchanged by this PR, so I'm going to leave it.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2521,25 +2521,27 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
-            }
+
+        // Only support compact block relay with witnesses
+        if (nCMPCTBLOCKVersion != CMPCTBLOCKS_VERSION) return;
```
`(pfrom.GetLocalServices() & NODE_WITNESS)` is always true (unless running on regtest with `-segwitheight=-1`). Perhaps this will be clearer after #21009 is merged.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -321,18 +321,12 @@ struct CNodeState {
-      * This is not used to gate request logic, as we really only care about fSupportsDesiredCmpctVersion,
-      * but is used as a flag to "lock in" the version of compact blocks (fWantsCmpctWitness) we send.
       */
     bool fProvidesHeaderAndIDs;
```
Yes, these parameters are currently confusing (hence this PR to clean them up!)

`fProvidersHeaderAndIDs` is currently used to make sure that the compact blocks version doesn't change after negotiation. If a peer sends `sendcmpct(version=2)`, then any future `sendcmpct` messages where version!=2 have no effect. The peer can change hb mode by sending `sendcmpct(version=2, hb_mode={true|false})`.

This is going away with this PR because the software will only recognize version 2 sendcmpct messages from now on.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2507,25 +2507,22 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+        nodestate->fPreferHeaderAndIDs = sendcmpct_hb;
+        // save whether peer selects us as BIP152 high-bandwidth peer
+        // (receiving sendcmpct(1) signals high-bandwidth, sendcmpct(0) low-bandwidth)
+        pfrom.m_bip152_highbandwidth_from = sendcmpct_hb;
```
This is unchanged by this PR. Eventually both of these fields should be moved to the `Peer` structure.

jnewbery commented:
Thanks for the review @jonatack @ariard @sipa @MarcoFalke . I think I've addresses all review comments now.

This PR is probably easier to review if rebased after #21009 is merged. That PR always sets `NODE_WITNESS` in local services, and so removes all the `pfrom->GetLocalServices() & NODE_WITNESS` calls in net_processing.

dhruv commented:
#21009 was split into #21009 and #21090. #21090 now sets `NODE_WITNESS` in local services, and so removes all the `pfrom->GetLocalServices() & NODE_WITNESS` calls in net_processing.

jnewbery commented:
Marking this as draft. It's a smaller and cleaner change once #21090 is merged, so let's wait for that to land first.

jnewbery commented:
I've rebased this on top of #21090. I'll still leave it as draft since we need that PR to be merged first, but this is ready for review if people are interested.

dhruv commented:
Reviewers interested in this PR, might also be interested in #21090 which is now ready for review.

jnewbery commented:
#21090 has been merged, so I've rebased this on master. This is now ready for review.

naumenkogs commented on test/functional/p2p_compactblocks.py:
```diff
@@ -306,8 +290,7 @@ def test_compactblock_construction(self, test_node, use_witness_address=True):
 
-        if use_witness_address:
-            assert segwit_tx_generated  # check that our test is not broken
+        assert segwit_tx_generated  # check that our test is not broken
```
nit: you could just do `assert not tx.wit.is_null()` above

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2713,17 +2713,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+            State(pfrom.GetId())->fWantsCmpctWitness = true;
         }
-        if (State(pfrom.GetId())->fWantsCmpctWitness == (nCMPCTBLOCKVersion == 2)) { // ignore later version announces
+        if (State(pfrom.GetId())->fWantsCmpctWitness) { // ignore later version announces
```
nit: does this inline-comment even make sense now?

naumenkogs commented:
ACK ef698b6eb3a684f52cbd2414ffc07893bbf69bbd

The PR seems ready to get rid of pre-segwit compact blocks, which makes sense now that segwit is used.

jnewbery commented:
fixed silent merge conflict in p2p_compactblocks_blocksonly.py

theuni commented on test/functional/p2p_compactblocks.py:
```diff
@@ -258,22 +251,14 @@ def check_announcement_of_new_block(node, peer, predicate):
+        test_node.send_and_ping(msg_sendcmpct(announce=False, version=2))
         check_announcement_of_new_block(node, test_node, lambda p: "cmpctblock" not in p.last_message and "headers" in p.last_message)
 
-        if old_node is not None:
```
If I'm reading correctly, I think we want to keep this test but reverse the expected outcome? That way we test that v1 doesn't work anymore as opposed to assuming it doesn't.

Disregard if I'm missing something else that exercises this.

jnewbery replied:
No, that's a good point. I've added the test in commit _[net processing] Only accept SENDCMPCT with version = 2_

theuni commented on src/net_processing.cpp:
```diff
@@ -2728,17 +2730,14 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-        if (pfrom.GetCommonVersion() >= SHORT_IDS_BLOCKS_VERSION) {
-            // Tell our peer we are willing to provide version 1 or 2 cmpctblocks
+        if (pfrom.GetCommonVersion() >= SHORT_IDS_BLOCKS_VERSION &&
+            WITH_LOCK(cs_main, {return State(pfrom.GetId())->fHaveWitness;})) {
```
Also, is this a behavioral change? It's not clear to me why the `fHaveWitness` check was added here.
Edit: Rather, why _wasn't_ it there before?

jnewbery replied:
I've removed the new check since it's not necessary.

theuni commented on src/net_processing.cpp:
```diff
@@ -2761,17 +2761,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+            State(pfrom.GetId())->fWantsCmpctWitness = true;
         }
-        if (State(pfrom.GetId())->fWantsCmpctWitness == (nCMPCTBLOCKVersion == 2)) { // ignore later version announces
+        if (State(pfrom.GetId())->fWantsCmpctWitness) { // ignore later version announces
```
Comment no longer makes sense here.

jnewbery replied:
removed!

theuni commented on src/net_processing.cpp:
```diff
@@ -2758,6 +2757,9 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
         LOCK(cs_main);
+        // Only support compact block relay with witness peers
+        if (!State(pfrom.GetId())->fHaveWitness) return;
```
According to BIP144, `NODE_WITNESS` means 

> A node will signal that it can provide witnesses using the following service bit

So I'm not sure this is strictly necessary. As a general theme though, I'm not sure bundling this logic with the value of NODE_WITNESS is necessary/correct.

jnewbery replied:
Agree. I've now removed it.

theuni reviewed (commented):
Concept ACK, code looks good, but I'm unsure about behavior.

I'm a little hazy on now `NODE_WITNESS` factors into these changes other than "one kinda seems to imply the other".

jnewbery commented:
Thanks for the review @theuni. I believe I've addressed all of your review comments.

> I'm a little hazy on now NODE_WITNESS factors into these changes other than "one kinda seems to imply the other".

You're right. I've removed those changes.

Also rebased on master.

theuni commented:
> > I'm a little hazy on now NODE_WITNESS factors into these changes other than "one kinda seems to imply the other".
> 
> You're right. I've removed those changes.

Thanks. To be clear though, these are seemingly reasonable changes worth evaluating separately. Just not necessary here for dropping v1 support.

theuni commented on src/net_processing.cpp:
```diff
@@ -2752,22 +2752,24 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         bool fAnnounceUsingCMPCTBLOCK = false;
         uint64_t nCMPCTBLOCKVersion = 0;
         vRecv >> fAnnounceUsingCMPCTBLOCK >> nCMPCTBLOCKVersion;
-        if (nCMPCTBLOCKVersion == 1 || nCMPCTBLOCKVersion == 2) {
```
Was there a reason for switching this to an early return, or is that now an artifact of old changes and rebases?

Just for readability, it's more clear imo wrapped in a `if` for the version check. I missed the early return when re-reviewing and was confused about why this change is safe.

jnewbery replied:
I think this is just personal taste. I _much_ prefer early returns where they're possible, especially compared with some of the very deeply nested if statements that we have in the code.

theuni replied:
This is nitty and not related to the code change, but I disagree with this. Deeply nested code, ugly as it may be, demonstrates the complexity of the logic in-place. Early returns don't simplify the logic, they just add sneaky trap-doors that aren't visible in diffs (In this case I misread the intent of the code because github didn't show enough context)

theuni commented on src/net_processing.cpp:
```diff
@@ -2760,17 +2760,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+            State(pfrom.GetId())->fWantsCmpctWitness = true;
         }
-        if (State(pfrom.GetId())->fWantsCmpctWitness == (nCMPCTBLOCKVersion == 2)) { // ignore later version announces
+        if (State(pfrom.GetId())->fWantsCmpctWitness) {
```
We could've previously fallen into this if `fWantsCmpctWitness == false && nCMPCTBLOCKVersion != 2`, but we won't now. I'm not sure what the consequence of that is.

jnewbery replied:
This if block is handling subsequent `sendcmpct` messages from the peer. The idea is that we'll ignore any subsequent `sendcmpct` message where the version doesn't match the version in the first received `sendcmpct` message, i.e. if we receive:

```
sendcmpct false 2
sendcmpct true 2
```

Then the first `sendcmpct` announces support for compact blocks and the second `sendcmpct` enables high-bandwidth mode. However, if we receive:

```
sendcmpct false 2
sendcmpct true 1
```

the second `sendcmpct` is ignored and we don't enable high-bandwidth mode.

This PR is removing all support for `sendcmcpt <true|false> 1`, so this specialized logic is no longer required.

theuni replied:
Yes, I see now. Thanks.

theuni commented on src/net_processing.cpp:
```diff
@@ -2752,22 +2752,24 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-            }
+
+        // Only support compact block relay with witnesses
+        if (nCMPCTBLOCKVersion != CMPCTBLOCKS_VERSION) return;
```
I believe this is a behavioral change in a pathological case:
```
SENDCMPCT true 2
SENDCMPCT false 1
```
Before this would have turned off announcements, but now it looks like it won't.
Again, I'm unsure of the consequence.

jnewbery replied:
> Before this would have turned off announcements, but now it looks like it won't.

I don't think this is true for the reason in https://github.com/bitcoin/bitcoin/pull/20799#discussion_r744939428. The second `sendcmpct` would not turn off announcements, since the if statement here:

```
            if (State(pfrom.GetId())->fWantsCmpctWitness == (nCMPCTBLOCKVersion == 2)) { // ignore later version announces
```

would not be true, so the logic to turn off hb announcements here:

```
                State(pfrom.GetId())->fPreferHeaderAndIDs = fAnnounceUsingCMPCTBLOCK;
```

would not be hit.

theuni replied:
I agree with this now, thanks for explaining!

theuni reviewed (commented):
Thanks for clarifying my other questions.

Is the commit "[net processing] Only relay blocks by cmpctblock and cache for fast relay if segwit is enabled" still supposed to be included here? Maybe it was intended to be dropped with the NODE_WITNESS changes?

I skimmed through it a few times and don't understand why it's ok, but I'm happy to take another pass at it if I've just missed something obvious.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1536,8 +1535,7 @@ void PeerManagerImpl::NewPoWValidBlock(const CBlockIndex *pindex, const std::sha
         CNodeState &state = *State(pnode->GetId());
         // If the peer has, or we announced to them the previous block already,
         // but we don't think they have this one, go ahead and announce it
-        if (state.m_sendcmpct_hb && (!fWitnessEnabled || state.m_sendcmpct) &&
```
On master, this line is:

```
        if (state.fPreferHeaderAndIDs && (!fWitnessEnabled || state.fWantsCmpctWitness) &&
```

i.e. if:
 - the peer requested that we relay hb compact blocks _and_
   - (the peer wants segwit-encoded (v2) compact blocks and this is a segwit block) _or_
   - (the peer wants non-segwit-encoded (v1) compact blocks and this block is from before segwit activation)

In early commits, we remove support for v1 compact blocks, so `fWantsCmpctWitness` is always true if `fPreferHeaderAndIDs` is true.

I've moved the changes around in the commits a bit, so hopefully this is a bit clearer now!

jnewbery commented:
Rebased and addressed @theuni's review comments.

jnewbery commented:
Rebased

dergoegge reviewed (commented):
Code review ACK 4c6d983708c94af10676048d1c1379be225b8f33

I think we could also remove the `bool fUseWTXID` param from the [`CBlockHeaderAndShortTxIDs` constructor](https://github.com/bitcoin/bitcoin/blob/7cc1860b12a51d6ff180da945d63d6aa03e3a4ca/src/blockencodings.h#L107) as it is now always set to true. What do you think?

jnewbery commented:
Thanks for the review @dergoegge!

> I think we could also remove the bool fUseWTXID param from the [CBlockHeaderAndShortTxIDs constructor](https://github.com/bitcoin/bitcoin/blob/7cc1860b12a51d6ff180da945d63d6aa03e3a4ca/src/blockencodings.h#L107) as it is now always set to true. What do you think?

I agree. I left it out of this PR since there are already lots of commits. I agree it should be done in a follow-up.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -2892,17 +2892,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         if (!State(pfrom.GetId())->fProvidesHeaderAndIDs) {
             State(pfrom.GetId())->fProvidesHeaderAndIDs = true;
-            State(pfrom.GetId())->fWantsCmpctWitness = nCMPCTBLOCKVersion == 2;
+            State(pfrom.GetId())->fWantsCmpctWitness = true;
```
nit: I don't see why 3223906fb7a50d089308c49c9f63e88982c515c2 is split up. In the line before you are ensuring that the version is only 2, but you leave this line untouched, which makes it confusing to review the previous commit.

Might be best to squash.

jnewbery replied:
I tried to split this PR into small commits after this comment: https://github.com/bitcoin/bitcoin/pull/20799#issuecomment-752197637. I'm not sure if there's a way to structure these changes that is going to satisfy everyone, so I'd prefer to leave this as is.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -367,12 +367,10 @@ struct CNodeState {
+    /** Whether this peer wants invs or cmpctblocks (when possible) for block announcements. */
+    bool m_requested_hb_cmpctblocks{false};
+    /** Whether this peer will send us cmpctblocks if we request them. */
+    bool m_provides_cmpctblocks{false};
```
744011062864d2d6dd03a5c8c381807d41063ac5: Any reason this isn't done as a scripted diff? The only other change seems to be the comment change, which I believe will be touched again anyway when this is moved to Peer.

jnewbery replied:
I tried to split this PR into small commits after this comment: https://github.com/bitcoin/bitcoin/pull/20799#issuecomment-752197637. I'm not sure if there's a way to structure these changes that is going to satisfy everyone, so I'd prefer to leave this as is.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -678,7 +678,6 @@ class PeerManagerImpl final : public PeerManager
     std::shared_ptr<const CBlock> m_most_recent_block GUARDED_BY(m_most_recent_block_mutex);
```
1aa6a5642327ee028d18b34e2c9b2be74c81454c commit message: s/is only to/is only used to/

?

jnewbery replied:
Updated commit text: s/that function is only to update/that function only updates/

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -3231,9 +3225,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-        CInv inv;
-        WITH_LOCK(cs_main, inv.type = State(pfrom.GetId())->m_provides_cmpctblocks ? MSG_WITNESS_BLOCK : MSG_BLOCK);
-        inv.hash = req.blockhash;
+        CInv inv(MSG_WITNESS_BLOCK, req.blockhash);
```
nit in 1aa6a5642327ee028d18b34e2c9b2be74c81454c:

CInv inv{...};

jnewbery replied:
Done

maflcko reviewed (approved):
ACK 4c6d983708c94af10676048d1c1379be225b8f33 🐉

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 4c6d983708c94af10676048d1c1379be225b8f33 🐉
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiu9Qv/eRGOPOxkDIxtTbbq7u/B5eVKZvLidzH3yNyz2rSFgVpLJyI8DcCdrxrL
b6sORYAtL428YdGe8YaVVMp+RHC/grTiP8jHPzL+dyFTZ9UNNL32bDMaFWyGY8Kd
/ppUcMPlGuSOmU4JPeTBuE/i8Sp0ii0u9kM4P1rYtGr2XKjJqhF7EOluFBuS3Pjf
WCYq6zDgLu6zcTLgdUJd9TEtEuOvQYa77LXJwZ3I7EJxUiDTLAflmgnoff9R/xiU
9+TTGV8bE1iTkRwwR86J4t2hpOlMe7e7tddCWzo/Jo6bwXdSMRpwlxrOjJ+3XrOo
l+iqrSewv9h2lajD+sdD8zj4R8ougXe61/Fgv6UbVA2olKM7VCFSCQdKOEBBNqtn
gMNJVoS6qm9bngN1kBJnDqmi8DFC//R29mWyvRHGDvy5dTT/CSqwkRK2mh+Cl2Pg
Q4EFJ0jVw5o07XumKei+iKSh0YgKgs6gzpfVI07aveP2i4E6zI7MZqbkkjvRDP1u
Ip3XHQrm
=/d5g
-----END PGP SIGNATURE-----
```


</details>

jnewbery commented on test/functional/p2p_compactblocks.py:
```diff
@@ -395,16 +370,12 @@ def check_compactblock_construction_from_block(self, version, header_and_shortid
         # request
         for announce in ["inv", "header"]:
-            block = self.build_block_on_tip(node, segwit=segwit)
+            block = self.build_block_on_tip(node, segwit=True)
```
That sounds like it could be done in a follow-up. This PR is already doing a lot.

I'm happy to review if you make this change.

ACONTOBIT commented:
@

jnewbery commented:
Thanks for the reviews @dergoegge and @MarcoFalke. I've addressed all review comments and rebased on latest master.

maflcko reviewed (approved):
re-ACK a50e34c367608fcec9697893981bfa294a7c08d after rebase 👓

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK a50e34c367608fcec9697893981bfa294a7c08d after rebase 👓
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgSIwwAx5jdanYqhTgoX2H5SG6pQ9aoBujOCn7R5sg8loLagrlzE9YOMkNxqTHJ
HY3V4lI3HFaa6TGqgowJNY2MbxNGQ3zqfepGsYe18NINnbnNDt6ZHG54uFQ65foV
hv4GRKRAZ4NGZ2mq28LVwLrVJTyQPIepurdlGLRmzSpRPcLnO81maL+48vFgle7T
H6o1sabgMWGffyfk6sUF+DxgYlk7JiKV8qVYcW3THAZhPuU1UBULB/fCZHnNjNEI
in5vTyIjEaraWyCpawc2OED0MjXghU1bwi8VU630hNE9DdnpBYMhq2TXKC52KM/n
SHIokmIIPk4mpif8S6RBZZGwUS5B28PgAoRfteEk5ce0P+CeewWnMppZFl7lMkSk
aHxl6e7UU7jqjtvotOE5TAubGMG0vn4oWnPx0k3PTtMMFRuSRj5gVvyH5D0y0BbL
90+0z3sS709purkQ+5q+zx11l1sMewFHTxcuVOyjMZ6aM2jd3SFi2MDfB/kPeyqd
mNXfG6pu
=E4BX
-----END PGP SIGNATURE-----
```


</details>

dergoegge commented:
ACK a50e34c367608fcec9697893981bfa294a7c08d1 - Only changes since my last review were a rebase and some nits.

Reviewed using `git range-diff 5d53cf38784df9ad9ed10306bf3fba3002fd9244..4c6d983708c94af10676048d1c1379be225b8f33 b74a6dde8cf50703a8814d402333580e4cdfea59..a50e34c367608fcec9697893981bfa294a7c08d1`

jnewbery commented:
Review suggestions implemented in #25147
